crates/optimism/flashblocks/src/service.rs

```diff
@@ -0,0 +1,179 @@
+    /// If the origin is the actual pending block, the block is built with withdrawals.
+    ///
+    /// After Cancun, if the origin is the actual pending block, the block includes the EIP-4788 pre
+    /// block contract call using the parent beacon block root received from the CL.
+    pub fn execute(&mut self) -> eyre::Result<ExecutedBlock<N>> {
+        let latest = self
+            .provider
+            .latest_header()?
+            .ok_or(EthApiError::HeaderNotFound(BlockNumberOrTag::Latest.into()))?;
+
+        let latest_attrs = self.builder.pending_env_attributes(&latest)?;
+
+        let state_provider = self.provider.history_by_block_hash(latest.hash())?;
+        let state = StateProviderDatabase::new(&state_provider);
+        let mut db = State::builder().with_database(state).with_bundle_update().build();
+
+        let mut builder = self
+            .evm_config
+            .builder_for_next_block(&mut db, &latest, latest_attrs)
+            .map_err(RethError::other)?;
+
+        builder.apply_pre_execution_changes()?;
+
+        let best_txs = self.blocks.iter().flat_map(|v| v.diff.transactions.clone());
+
+        for tx in best_txs {
+            let tx = N::SignedTx::decode(&mut tx.as_ref())?;
+            let signer = tx.recover_signer_unchecked()?;
+            let tx = Recovered::new_unchecked(tx, signer);
+            let _gas_used = builder.execute_transaction(tx)?;
+        }
+
+        let BlockBuilderOutcome { execution_result, block, hashed_state, .. } =
+            builder.finish(&state_provider)?;
+
+        let execution_outcome = ExecutionOutcome::new(
+            db.take_bundle(),
+            vec![execution_result.receipts],
+            block.number(),
+            vec![execution_result.requests],
+        );
+
+        Ok(ExecutedBlock {
+            recovered_block: block.into(),
+            execution_output: Arc::new(execution_outcome),
+            hashed_state: Arc::new(hashed_state),
+        })
+    }
+}
+
+impl<
+        N: NodePrimitives,
+        S: Stream<Item = FlashBlock> + Unpin,
+        EvmConfig: ConfigureEvm<Primitives = N, NextBlockEnvCtx: Unpin>,
+        Provider: StateProviderFactory
+            + BlockReaderIdExt<
+                Header = HeaderTy<N>,
+                Block = BlockTy<N>,
+                Transaction = N::SignedTx,
+                Receipt = ReceiptTy<N>,
+            > + Unpin,
+        Builder: PendingEnvBuilder<EvmConfig>,
+    > Stream for FlashBlockService<N, S, EvmConfig, Provider, Builder>
+{
+    type Item = eyre::Result<ExecutedBlock<N>>;
+
+    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+        let this = self.get_mut();
+        loop {
+            match this.rx.poll_next_unpin(cx) {
+                Poll::Ready(Some(flashblock)) => this.add_flash_block(flashblock),
+                Poll::Ready(None) => return Poll::Ready(None),
+                Poll::Pending => return Poll::Ready(Some(this.execute())),
```

Comment:
how this part will actually work we'll need to figure out we actually wire this to the rpc part, so make this doesn't need to be a stream entirely and could manage the pending block internally entirely

but we can figure this part out later